### ISSUE TITLE
beforeEach does not like supertest

### ISSUE THREAD

**Issue by pke** [tapjs/tapjs#596]
When using a `beforeEach` hook the following test is never executed.

Moving the `supertest` request creation inside the test works though.

```js
const { test } = require("tap")
const supertest = require("supertest")

const { app } = require("../src/server")

test("server tests", t => {
  const server = app.listen()

  t.tearDown(() => server.close())

  /*t.beforeEach(() => {
    t.context.request = supertest(server)
  })*/

  t.test("POST a URL to shorten it", t => {
    const request = supertest(server)
    request
      .post("/https://www.google.com?q=whats%2042")
      .expect(201)
      .end((err, body) => {
        if (err) {
          throw err
        } 
        t.is(body.text, `http://127.0.0.1:${server.address().port}/a18a36`)
        t.end()
      })
  })

  t.end()
})
``

**Comment by isaacs** [comment]
`t.beforeEach` needs to either call the `done` callback provided, or return a Promise.

Does this work?

```js
  t.beforeEach(done => {
    t.context.request = supertest(server)
    done()
  })
```

**Comment by pke** [comment]
Silly me. I was thinking `done` is so "pre-promise time" that the API docs were surely outdated and that when its not called, tap will treat the before function as being done.
Whats the (historical) technical reason behind the `done` function to be called? 
Could the API changed to make this call not mandatory?

**Comment by isaacs** [comment]
Simplest way is to just make it an async function.

This  feature is copied from mocha.  In mocha, the `beforeEach` function inspects the `function.length` property to determine whether it expects an argument, and if not, and it doesn't return a promise, then it presumes it to be sync.  That seemed wildly unsafe to me, so I made it take a callback all the time.  Then Promises became a thing, so I made it respond to a returned promise.

The `t.teardown()` function is the opposite.  It started out as a simple event listener, so it was always synchronous.  Then, some time around v14 I believe, I made it respond to a returned Promise.

I've contemplated making `beforeEach`  and `afterEach` sync-unless-promisey as well, but that's a breaking change that'll cause a lot of tests in the wild to break, and it's rare that people get bitten by it as it is today, so I've shied away from making that change.
